src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SingleInitNullable.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using Roslyn.Utilities;
+
+/// <summary>
+/// A lazily initialized version of <see cref="Nullable{T}"/> which uses the same space as a <see cref="Nullable{T}"/>.
+/// </summary>
+[NonCopyable]
+internal struct SingleInitNullable<T>
+    where T : struct
+{
+    /// <summary>
+    /// One of three values:
+    /// <list type="bullet">
+    /// <item>0. <see cref="_value"/> is not initialized yet.</item>
+    /// <item>1. <see cref="_value"/> is currently being initialized by some thread.</item>
+    /// <item>2. <see cref="_value"/> has been initialized.</item>
+    /// </list>
+    /// </summary>
+    private int _initialized;
+
+    /// <summary>
+    /// Actual stored value.  Only safe to read once <see cref="_initialized"/> is set to 2.
+    /// </summary>
+    private T _value;
+
+    /// <summary>
+    /// Ensure that the given target value is initialized in a thread-safe manner.
+    /// </summary>
+    /// <param name="valueFactory">A factory delegate to create a new instance of the target value. Note that this
+    /// delegate may be called more than once by multiple threads, but only one of those values will successfully be
+    /// written to the target.</param>
+    /// <returns>The target value.</returns>
+    public T Initialize<TArg>(Func<TArg, T> valueFactory, TArg arg)
+        => ReadIfInitialized() ?? GetOrStore(valueFactory(arg));
+
+    private T? ReadIfInitialized()
+        => Volatile.Read(ref _initialized) == 2 ? _value : null;
+
+    private T GetOrStore(T value)
+    {
+        SpinWait spinWait = default;
+        while (true)
+        {
+            switch (Interlocked.CompareExchange(ref _initialized, 1, 0))
```

Comment:
Nit: use named parameters here, since I've absolutely seen bugs where people cross the zero and one and bad things happen. :smile:

Reply:
sure.